File: 2698_find_the_punishment_number_of_an_integer.cpp
class Solution {
public:
    int punishmentNumber(int n) {
        int z=0;
        for(int i=1;i<=n;i++) {
            std::string s = std::to_string(i*i);
            if(h(s,0,i,0)) z+=i*i;
        }
        return z;
    }
private:
    bool h(const std::string &s, int x, int y, int c) {
        if(x==s.size()) return c==y;
        int p=0;
        for(int i=x;i<s.size();i++) {
            p=p*10+(s[i]-'0');
            if(c+p>y) break;
            if(h(s,i+1,y,c+p)) return true;
        }
        return false;
    }
};

File: 2698_find_the_punishment_number_of_an_integer.java
class Solution {
    public int punishmentNumber(int n) {
        int w = 0;
        for (int i = 1; i <= n; i++) {
            String s = String.valueOf(i * i);
            if (m(s, 0, i, 0)) w += i * i;
        }
        return w;
    }

    private boolean m(String s, int x, int y, int c) {
        if (x == s.length()) return c == y;
        int t = 0;
        for (int i = x; i < s.length(); i++) {
            t = t * 10 + (s.charAt(i) - '0');
            if (c + t > y) break;
            if (m(s, i + 1, y, c + t)) return true;
        }
        return false;
    }
}
